src/jit/MemoryInl.h

```diff
@@ -1011,3 +1011,326 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
+static void atomicRmwAdd64(uint64_t* shared_p, int64_t* value, int64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->fetch_add(*value);
+}
+
+static void atomicRmwSub64(uint64_t* shared_p, int64_t* value, int64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->fetch_sub(*value);
+}
+
+static void atomicRmwAnd64(uint64_t* shared_p, int64_t* value, int64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->fetch_and(*value);
+}
+
+static void atomicRmwOr64(uint64_t* shared_p, int64_t* value, int64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->fetch_or(*value);
+}
+
+static void atomicRmwXor64(uint64_t* shared_p, int64_t* value, int64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->fetch_xor(*value);
+}
+
+static void atomicRmwXchg64(uint64_t* shared_p, int64_t* value, int64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->exchange(*value);
+}
+
+static void emitAtomicRmw64(sljit_compiler* compiler, Instruction* instr)
+{
+    uint32_t options = MemAddress::CheckNaturalAlignment | MemAddress::AbsoluteAddress;
+    uint32_t size = 8;
+    sljit_u32 offset;
+
+    Operand* operands = instr->operands();
+    MemAddress addr(options, instr->requiredReg(0), instr->requiredReg(1), 0);
+    AtomicRmw* rmwOperation = reinterpret_cast<AtomicRmw*>(instr->byteCode());
+    offset = rmwOperation->offset();
+
+    addr.check(compiler, operands, offset, size);
+
+    if (addr.memArg.arg == 0) {
+        return;
+    }
+
+    JITArgPair srcArgPair(operands + 1);
+    JITArgPair dstArgPair(operands + 2);
+    sljit_s32 type = SLJIT_ARGS3V(P, P, P);
+    sljit_s32 faddr;
+
+    switch (instr->opcode()) {
+    case ByteCode::I64AtomicRmwAddOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwAdd64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwSubOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwSub64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwAndOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwAnd64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwOrOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwOr64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwXorOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwXor64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwXchgOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwXchg64);
+        break;
+    }
+    default: {
+        RELEASE_ASSERT_NOT_REACHED();
+        break;
+    }
+    }
+
+    if (srcArgPair.arg1 != SLJIT_MEM1(kFrameReg)) {
```

Comment:
Can this be combined with `if (srcArgPair.arg1 == SLJIT_MEM1(kFrameReg)) {` below?